### DISCUSSION TITLE
How to run a "For-Each" loop over the controls of a UF

### DISCUSSION THREAD

**Discussion by Peter9Claudius** [twinbasic/twinbasic#990] [Q&A] [answered]
Greetings everyone, I am impressed with the Visual Basic-like syntax in this development environment. Have also tried this and that successfully, but have stumbled over a hurdle: A For-Each loop to iterate the controls on a UF. With every variation I have tried, I always get the same error: "Run-time error -2147467259 (80004005) - unknown error". No explanation, the pointer always on the first line:
 
   Dim ctl As Object
    
   _For Each ctl In Me.Controls_ '<= unknown error
        Debug.Print ctl.Name
    Next ctl

This is how I usually do it in VBA. Can anyone help me with advice or even give me an example?

Thank you very much and best regards,

Peter.

**Comment by rexxitall** [comment]
should work usually but maybe not yet implemented. if you get something useful with debug.print me.controls.count then i would asume that a for next loop will work. for i=0 to me.controls.count-1: debug.print me.controls.item(i).name: next. Just a thought for a workaround.

**Reply by Peter9Claudius** [reply]
Hello, thanks for the tip. Unfortunately, it does not work that way. This time the error "Run-time error 91 (0000005B) /Description: Object variable or with block variable not set" is issued and the pointer is again set to the first line of this loop:
           > For i = 0 To Me.Controls.count - 1
            Debug.Print Me.Controls.item(i).name
            Next
This loop is placed in the code of the form, but "Form1.Controls.count" is apparently not recognized either, although the object name is "Form1". The Intellisense window suggests that only "Controls" is known for "Me" or "Form1", but neither "item" nor "count". 
Furthermore, the For-Each loop itself seems to be known, because the following array pass brings valid results:
   Dim strWord(3) As Variant, itm As Variant
   strWord(0) = "Adam": strWord(1) = "Eve": strWord(2) = "Snake"
       For Each itm In strWord
             Debug.Print itm
       Next itm
> Snake
    Eve
    Adam
The fact that iteration and casting of objects via "controls" is not possible in the form is exceedingly annoying.

**Comment by WaynePhillipsEA** [comment, marked as answer]
The Controls property is not yet populated, it returns Nothing.  I will open an issue for tracking support for it.   Thanks.

**Reply by Peter9Claudius** [reply]
Thank you very much! :-)

**Reply by Peter9Claudius** [reply]
Hi, I wanted to give a quick feedback after pulling the latest release. The error is still there exactly the same. No iteration via "controls" possible. Kind regards, Peter.

**Reply by WaynePhillipsEA** [reply]
Please refer to #992 which was created for tracking it, and will be updated once a release addresses it.  Each release includes notes to let you know what was fixed/changed (e.g. #994).

**Reply by Peter9Claudius** [reply]
Many, many **thanks**! This looks just great with the current release! **Perfect**! :-)
